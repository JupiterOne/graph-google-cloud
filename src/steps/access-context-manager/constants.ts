export const STEP_ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES =
  'fetch-access-context-manager-access-policies';
export const STEP_ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS =
  'fetch-access-context-manager-access-levels';
export const STEP_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS =
  'fetch-access-context-manager-service-perimeters';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY = 'AccessPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY =
  'google_access_context_manager_access_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL = 'Ruleset';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL =
  'google_access_context_manager_access_level';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER =
  'google_access_context_manager_service_perimeter';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_EGRESS_POLICY =
  'ControlPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_EGRESS_POLICY =
  'google_access_context_manager_service_perimeter_egress_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_INGRESS_POLICY =
  'ControlPolicy';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_INGRESS_POLICY =
  'google_access_context_manager_service_perimeter_ingress_policy';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_API_OPERATION =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_API_OPERATION =
  'google_access_context_manager_service_perimeter_api_operation';

export const ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_METHOD_SELECTOR =
  'Configuration';
export const ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER_METHOD_SELECTOR =
  'google_access_context_manager_service_perimeter_method_selector';

export const RELATIONSHIP_TYPE_ACCESS_POLICY_HAS_ACCESS_LEVEL =
  'google_access_context_manager_access_policy_has_level';
export const RELATIONSHIP_TYPE_ACCESS_POLICY_HAS_SERVICE_PERIMETER =
  'google_access_context_manager_access_policy_has_service_perimeter';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_PROTECTS_PROJECT =
  'google_access_context_manager_service_perimeter_protects_project';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_LIMITS_API_SERVICES =
  'google_access_context_manager_service_perimeter_limits_api_service';

export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_HAS_EGRESS_POLICY =
  'google_access_context_manager_service_perimeter_has_egress_policy';
export const RELATIONSHIP_TYPE_EGRESS_POLICY_HAS_API_OPERATION =
  'google_access_context_manager_service_perimeter_egress_policy_has_api_operation';
export const RELATIONSHIP_TYPE_API_OPERATION_HAS_METHOD_SELECTOR =
  'google_access_context_manager_service_perimeter_api_operation_has_method_selector';
export const RELATIONSHIP_TYPE_SERVICE_PERIMETER_HAS_INGRESS_POLICY =
  'google_access_context_manager_service_perimeter_has_ingress_policy';
export const RELATIONSHIP_TYPE_INGRESS_POLICY_HAS_API_OPERATION =
  'google_access_context_manager_service_perimeter_ingress_policy_has_api_operation';

export const IngestionSources = {
  ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES: 'acm-access-policies',
  ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS: 'acm-access-levels',
  ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS: 'acm-service-perimeters',
};

export const AccessContextManagerIngestionConfig = {
  [IngestionSources.ACCESS_CONTEXT_MANAGER_ACCESS_POLICIES]: {
    title: 'Google Cloud Access Context Manager Access Policies',
    description: 'Defines secure access boundaries in GCP.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ACCESS_CONTEXT_MANAGER_ACCESS_LEVELS]: {
    title: 'Google Cloud Access Context Manager Access Levels',
    description: 'Manages access via hierarchical levels in GCP.',
    defaultsToDisabled: false,
  },
  [IngestionSources.ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETERS]: {
    title: 'Google Cloud Access Context Manager Service Perimeters',
    description: 'Secures resources within network boundaries.',
    defaultsToDisabled: false,
  },
};
