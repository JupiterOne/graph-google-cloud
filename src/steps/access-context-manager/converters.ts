import { accesscontextmanager_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import {
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
  ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
  ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
} from './constants';

export function createAccessPolicyEntity(
  data: accesscontextmanager_v1.Schema$AccessPolicy,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_POLICY,
        name: data.name,
        title: data.title,
        etag: data.etag,
      },
    },
  });
}

export function createAccessLevelEntity(
  data: accesscontextmanager_v1.Schema$AccessLevel,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_ACCESS_LEVEL,
        name: data.name,
        description: data.description,
        title: data.title,
      },
    },
  });
}

export function createServicePerimeterEntity(
  data: accesscontextmanager_v1.Schema$ServicePerimeter,
) {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _key: data.name as string,
        _type: ENTITY_TYPE_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
        _class: ENTITY_CLASS_ACCESS_CONTEXT_MANAGER_SERVICE_PERIMETER,
        name: data.name,
        description: data.description,
        perimeterType: data.perimeterType,
        title: data.title,
        useExplicitDryRunSpec: data.useExplicitDryRunSpec,
        // shared fields between .spec/.status depending on the run mode
        accessLevels: data.status?.accessLevels,
        resources: data.status?.resources,
        restrictedServices: data.status?.restrictedServices,
        // ingressPolicies
        // egressPolicies
      },
    },
  });
}
