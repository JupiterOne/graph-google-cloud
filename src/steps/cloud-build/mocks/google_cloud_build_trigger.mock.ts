export default {
  id: '0f5d9ec8-f7d7-4067-8912-baf2133be725',
  createTime: '2022-08-24T22:09:57.281610382Z',
  name: 'testing',
  gitFileSource: {
    path: 'cloudbuild.yaml',
    uri: 'https://github.com/sample-user/sample-repo',
    repoType: 'GITHUB',
    revision: 'refs/heads/master',
  },
  sourceToBuild: {
    uri: 'https://github.com/sample-user/sample-repo',
    ref: 'refs/heads/master',
    repoType: 'GITHUB',
  },
};
