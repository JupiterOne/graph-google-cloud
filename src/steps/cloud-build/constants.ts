import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { CloudSourceRepositoriesEntitiesSpec } from '../cloud-source-repositories/constants';
import { StorageEntitiesSpec } from '../storage/constants';

export const CloudBuildEntitiesSpec = {
  BUILD: {
    resourceName: 'Cloud Build',
    _type: 'google_cloud_build',
    _class: ['Workflow'],
  },
  BUILD_TRIGGER: {
    resourceName: 'Cloud Build Trigger',
    _type: 'google_cloud_build_trigger',
    _class: ['Rule'],
  },
  BUILD_WORKER_POOL: {
    resourceName: 'Cloud Build Worker Pool',
    _type: 'google_cloud_build_worker_pool',
    _class: ['Cluster'],
  },
  BUILD_GITHUB_ENTERPRISE_CONFIG: {
    resourceName: 'Cloud Build GitHub Enterprise Config',
    _type: 'google_cloud_github_enterprise_config',
    _class: ['Configuration'],
  },
  BUILD_BITBUCKET_SERVER_CONFIG: {
    resourceName: 'Cloud Build BitBucket Server Config',
    _type: 'google_cloud_bitbucket_server_config',
    _class: ['Configuration'],
  },
  BUILD_BITBUCKET_REPO: {
    resourceName: 'Cloud Build BitBucket Server Repo',
    _type: 'google_cloud_bitbucket_server_repo',
    _class: ['CodeRepo'],
  },
};

export const CloudBuildRelationshipsSpec = {
  BITBUCKET_SERVER_HAS_REPO: {
    _type: 'google_cloud_bitbucket_server_config_has_repo',
    sourceType: CloudBuildEntitiesSpec.BUILD_BITBUCKET_SERVER_CONFIG._type,
    _class: RelationshipClass.HAS,
    targetType: CloudBuildEntitiesSpec.BUILD_BITBUCKET_REPO._type,
  },
  BUILD_TRIGGER_TRIGGERS_BUILD: {
    _type: 'google_cloud_build_trigger_triggers_build',
    sourceType: CloudBuildEntitiesSpec.BUILD_TRIGGER._type,
    _class: RelationshipClass.TRIGGERS,
    targetType: CloudBuildEntitiesSpec.BUILD._type,
  },
  BUILD_USES_STORAGE_BUCKET: {
    _type: 'google_cloud_build_uses_storage_bucket',
    _class: RelationshipClass.USES,
    sourceType: CloudBuildEntitiesSpec.BUILD._type,
    targetType: StorageEntitiesSpec.STORAGE_BUCKET._type,
  },
  BUILD_USES_SOURCE_REPOSITORY: {
    _type: 'google_cloud_build_uses_source_repository',
    _class: RelationshipClass.USES,
    sourceType: CloudBuildEntitiesSpec.BUILD._type,
    targetType: CloudSourceRepositoriesEntitiesSpec.REPOSITORY._type,
  },
};

export const CloudBuildStepsSpec = {
  FETCH_BUILDS: {
    id: 'fetch-cloud-builds',
    name: 'Fetch Cloud Builds',
  },
  FETCH_BUILD_TRIGGERS: {
    id: 'fetch-cloud-build-triggers',
    name: 'Fetch Cloud Build Triggers',
  },
  FETCH_BUILD_WORKER_POOLS: {
    id: 'fetch-cloud-build-worker-pools',
    name: 'Fetch Cloud Build Worker Pools',
  },
  FETCH_BUILD_GITHUB_ENTERPRISE_CONFIG: {
    id: 'fetch-cloud-build-ghe-configs',
    name: 'Fetch Cloud Build GitHub Enterprise Configs',
  },
  FETCH_BUILD_BITBUCKET_SERVER_CONFIG: {
    id: 'fetch-cloud-build-bb-configs',
    name: 'Fetch Cloud Build BitBucket Server Configs',
  },
  FETCH_BUILD_BITBUCKET_REPOS: {
    id: 'fetch-cloud-build-bb-repos',
    name: 'Fetch Cloud Build BitBucket Server Repos',
  },
  BUILD_CLOUD_BUILD_TRIGGER_TRIGGERS_BUILD_RELATIONSHIPS: {
    id: 'build-cloud-build-trigger-triggers-build-relationships',
    name: 'Build Cloud Build Trigger -> Cloud Build Relationships',
  },
  BUILD_CLOUD_BUILD_USES_STORAGE_BUCKET_RELATIONSHIPS: {
    id: 'build-cloud-build-uses-storage-bucket-relationships',
    name: 'Build Cloud Build -> Storage Bucket Relationships',
  },
  BUILD_CLOUD_BUILD_USES_SOURCE_REPOSITORY_RELATIONSHIPS: {
    id: 'build-cloud-build-uses-source-repo-relationships',
    name: 'Build Cloud Build -> Source Repository Relationships',
  },
  BUILD_CLOUD_BUILD_TRIGGER_USES_GITHUB_REPO_RELATIONSHIPS: {
    id: 'build-cloud-build-trigger-uses-github-repo',
    name: 'Build Cloud Build Trigger -> Github Repository Relationships',
  },
};

// https://cloud.google.com/build/docs/locations
export const CloudBuildLocations = [
  'asia-south1',
  'asia-south2',
  'asia-southeast1',
  'asia-southeast2',
  'australia-southeast1',
  'australia-southeast2',
  'europe-central2',
  'europe-north1',
  'europe-west1',
  'europe-west2',
  'europe-west3',
  'europe-west4',
  'europe-west6',
  'northamerica-northeast1',
  'northamerica-northeast2',
  'southamerica-east1',
  'southamerica-west1',
  'us-central1',
  'us-east1',
  'us-east4',
  'us-west1',
  'us-west2',
  'us-west3',
  'us-west4',
];
