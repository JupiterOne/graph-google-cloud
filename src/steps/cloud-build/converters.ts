import { parseTimePropertyValue } from '@jupiterone/integration-sdk-core';
import { cloudbuild_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../utils/entity';
import { CloudBuildEntitiesSpec } from './constants';

export const createGoogleCloudBuildEntity = (
  source: cloudbuild_v1.Schema$Build,
) => {
  const data = {
    ...source,
    labels: source.tags,
    tags: undefined,
  };

  return createGoogleCloudIntegrationEntity(source, {
    entityData: {
      source: source,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD._class,
        _type: CloudBuildEntitiesSpec.BUILD._type,
        _key: data.id as string,
        name: data.id as string,
        id: data.id as string,
        status: data.status,
        'source.storageSource.bucket': data.source?.storageSource?.bucket,
        'source.storageSource.object': data.source?.storageSource?.object,
        'source.storageSource.generation':
          data.source?.storageSource?.generation,
        'results.buildStepImages': data.results?.buildStepImages,
        'results.buildStepOutputs': data.results?.buildStepOutputs,
        createTime: parseTimePropertyValue(data.createTime),
        startTime: parseTimePropertyValue(data.startTime),
        finishTime: parseTimePropertyValue(data.finishTime),
        timeout: data.timeout,
        projectId: data.projectId,
        logsBucket: data.logsBucket,
        'options.logging': data.options?.logging,
        'options.pool': data.options?.workerPool,
        logUrl: data.logUrl,
        queueTtl: data.queueTtl,
      },
    },
  });
};

export const createGoogleCloudBuildTriggerEntity = (
  source: cloudbuild_v1.Schema$BuildTrigger,
) => {
  const data = {
    ...source,
    labels: source.tags,
    tags: undefined,
  };

  return createGoogleCloudIntegrationEntity(source, {
    entityData: {
      source: source,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_TRIGGER._class,
        _type: CloudBuildEntitiesSpec.BUILD_TRIGGER._type,
        _key: data.id as string,
        name: data.name,
        id: data.id as string,
        createdOn: parseTimePropertyValue(source.createTime),
      },
    },
  });
};

export const createGoogleCloudBuildWorkerPoolEntity = (
  data: cloudbuild_v1.Schema$WorkerPool,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: {
        ...data,
        tags: data.annotations,
      },
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_WORKER_POOL._class,
        _type: CloudBuildEntitiesSpec.BUILD_WORKER_POOL._type,
        _key: data.uid as string,
        name: data.name,
        displayName: data.displayName as string,
        createdOn: parseTimePropertyValue(data.createTime),
        updatedOn: parseTimePropertyValue(data.updateTime),
        deletedOn: parseTimePropertyValue(data.deleteTime),
        state: data.state,
        uid: data.uid,
      },
    },
  });
};

export const createGoogleCloudBuildGithubEnterpriseConfigEntity = (
  data: cloudbuild_v1.Schema$GitHubEnterpriseConfig,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_GITHUB_ENTERPRISE_CONFIG._class,
        _type: CloudBuildEntitiesSpec.BUILD_GITHUB_ENTERPRISE_CONFIG._type,
        _key: data.name as string,
        name: data.name,
        displayName: data.displayName as string,
        appId: data.appId,
        createdOn: parseTimePropertyValue(data.createTime),
        hostUrl: data.hostUrl,
        peeredNetwork: data.peeredNetwork,
        'secrets.oauthClientIdVersionName':
          data.secrets?.oauthClientIdVersionName,
        'secrets.oauthSecretVersionName': data.secrets?.oauthSecretVersionName,
        'secrets.privateKeyVersionName': data.secrets?.privateKeyVersionName,
        'secrets.webhookSecretVersionName':
          data.secrets?.webhookSecretVersionName,
        sslCa: data.sslCa,
        webhookKey: data.webhookKey,
      },
    },
  });
};

export const createGoogleCloudBuildBitbucketServerConfigEntity = (
  data: cloudbuild_v1.Schema$BitbucketServerConfig,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_BITBUCKET_SERVER_CONFIG._class,
        _type: CloudBuildEntitiesSpec.BUILD_BITBUCKET_SERVER_CONFIG._type,
        _key: data.name as string,
        name: data.name,
        hostUri: data.hostUri,
        createTime: parseTimePropertyValue(data.createTime),
        username: data.username,
        webhookKey: data.webhookKey,
        apiKey: data.apiKey,
        'secrets.adminAccessTokenVersionName':
          data?.secrets?.adminAccessTokenVersionName,
        'secrets.readAccessTokenVersionName':
          data?.secrets?.readAccessTokenVersionName,
        'secrets.webhookSecretVersionName':
          data?.secrets?.webhookSecretVersionName,
      },
    },
  });
};

export const createGoogleCloudBuildBitbucketRepoEntity = (
  data: cloudbuild_v1.Schema$BitbucketServerRepository,
) => {
  return createGoogleCloudIntegrationEntity(data, {
    entityData: {
      source: data,
      assign: {
        _class: CloudBuildEntitiesSpec.BUILD_BITBUCKET_REPO._class,
        _type: CloudBuildEntitiesSpec.BUILD_BITBUCKET_REPO._type,
        _key: data.name as string,
        name: data.name,
        browseUri: data.browseUri,
        description: data.description,
        displayName: data.displayName as string,
        'repoId.projectKey': data.repoId?.projectKey,
        'repoId.repoSlug': data.repoId?.repoSlug,
        'repoId.webhookId': data.repoId?.webhookId,
      },
    },
  });
};
