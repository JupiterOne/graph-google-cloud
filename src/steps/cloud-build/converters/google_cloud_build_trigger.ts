import { cloudbuild_v1 } from 'googleapis';
import { createGoogleCloudIntegrationEntity } from '../../../utils/entity';
import { CloudBuildEntitiesSpec } from '../constants';

export default (source: cloudbuild_v1.Schema$BuildTrigger) => {
  const data = {
    ...source,
    labels: source.tags,
    tags: undefined,
  };

  return createGoogleCloudIntegrationEntity(source, {
    entityData: {
      source: source,
      assign: {
        _class: CloudBuildEntitiesSpec.TRIGGER._class,
        _type: CloudBuildEntitiesSpec.TRIGGER._type,
        _key: data.id as string,
        name: data.name,
        id: data.id as string,
      },
    },
  });
};
