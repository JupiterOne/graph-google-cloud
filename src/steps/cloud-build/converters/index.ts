import google_cloud_build from './google_cloud_build';
import google_cloud_build_trigger from './google_cloud_build_trigger';

export default {
  google_cloud_build: google_cloud_build,
  google_cloud_build_trigger: google_cloud_build_trigger,
};
