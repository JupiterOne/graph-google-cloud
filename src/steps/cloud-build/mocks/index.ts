import google_cloud_build from './google_cloud_build.mock';
import google_cloud_build_trigger from './google_cloud_build_trigger.mock';

export default { google_cloud_build, google_cloud_build_trigger };
