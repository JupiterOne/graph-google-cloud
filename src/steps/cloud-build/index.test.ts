import {
  executeStepWithDependencies,
  Recording,
  StepTestConfig,
} from '@jupiterone/integration-sdk-testing';
import { invocationConfig } from '../..';
import { integrationConfig } from '../../../test/config';
import {
  getMatchRequestsBy,
  setupGoogleCloudRecording,
} from '../../../test/recording';
import { CloudBuildEntitiesSpec, CloudBuildStepsSpec } from './constants';
import converters from './converters';
import mocks from './mocks';

describe('#cloud-build', () => {
  describe('steps', () => {
    let recording: Recording;
    afterEach(async () => {
      if (recording) await recording.stop();
    });

    for (const step of Object.values(CloudBuildStepsSpec)) {
      test(step.id, async () => {
        recording = setupGoogleCloudRecording({
          name: step.id,
          directory: __dirname,
          options: {
            matchRequestsBy: getMatchRequestsBy(integrationConfig),
          },
        });

        const stepTestConfig: StepTestConfig = {
          stepId: step.id,
          instanceConfig: integrationConfig,
          invocationConfig: invocationConfig as any,
        };

        const result = await executeStepWithDependencies(stepTestConfig);
        expect(result).toMatchStepMetadata(stepTestConfig);
      });
    }
  });

  describe('entities', () => {
    for (const entity of Object.values(CloudBuildEntitiesSpec)) {
      test(entity._type, () => {
        expect(converters[entity._type](mocks[entity._type])).toMatchSnapshot();
      });
    }
  });
});
