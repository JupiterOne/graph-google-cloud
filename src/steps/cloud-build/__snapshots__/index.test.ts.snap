// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#cloud-build entities google_cloud_build 1`] = `
Object {
  "_class": Array [
    "Secret",
  ],
  "_key": "projects/167984947943/locations/global/builds/58991a72-f158-481d-ad66-771b90ce3e65",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "createTime": "2022-08-17T20:44:37.573002899Z",
        "failureInfo": Object {
          "detail": "Build step failure: build step 0 \\"gcr.io/cloud-builders/gsutil\\" failed: step exited with non-zero status: 1",
          "type": "USER_BUILD_STEP",
        },
        "finishTime": "2022-08-17T20:44:46.243182675Z",
        "id": "58991a72-f158-481d-ad66-771b90ce3e65",
        "logUrl": "https://console.cloud.google.com/cloud-build/builds/58991a72-f158-481d-ad66-771b90ce3e65?project=167984947943",
        "logsBucket": "gs://167984947943.cloudbuild-logs.googleusercontent.com",
        "name": "projects/167984947943/locations/global/builds/58991a72-f158-481d-ad66-771b90ce3e65",
        "options": Object {
          "logging": "LEGACY",
          "pool": Object {},
        },
        "projectId": "j1-gc-integration-dev-v3",
        "queueTtl": "3600s",
        "results": Object {
          "buildStepImages": Array [
            "sha256:0229d2fde1be1a278c6bb73fbfd98b2df63c30cb9012c882bdfa228e738afdf5",
            "",
            "",
            "",
            "",
          ],
          "buildStepOutputs": Array [
            "",
            "",
            "",
            "",
            "",
          ],
        },
        "source": Object {
          "storageSource": Object {
            "bucket": "j1-gc-integration-dev-v3_cloudbuild",
            "generation": "1660769077096260",
            "object": "source/1660769074.865182-d160c8803b704c60b50e68d2b3b97d73.tgz",
          },
        },
        "sourceProvenance": Object {
          "fileHashes": Object {
            "gs://j1-gc-integration-dev-v3_cloudbuild/source/1660769074.865182-d160c8803b704c60b50e68d2b3b97d73.tgz#1660769077096260": Object {
              "fileHash": Array [
                Object {
                  "type": "MD5",
                  "value": "eg9jmTQDqknuC0oat864jA==",
                },
              ],
            },
          },
          "resolvedStorageSource": Object {
            "bucket": "j1-gc-integration-dev-v3_cloudbuild",
            "generation": "1660769077096260",
            "object": "source/1660769074.865182-d160c8803b704c60b50e68d2b3b97d73.tgz",
          },
        },
        "startTime": "2022-08-17T20:44:38.421413511Z",
        "status": "FAILURE",
        "steps": Array [
          Object {
            "args": Array [
              "-c",
              "gsutil ls -b gs://<some bucket for caching gradle file> || gsutil mb -l gs://<some bucket for caching gradle file>

(
  gsutil cp gs://<some bucket for caching gradle file>/<file to hold cached content>.tar.gz /tmp/gradle-cache.tar.gz &&
  tar -xzf /tmp/gradle-cache.tar.gz
) || echo 'Cache not found'
",
            ],
            "dir": "/cachinghome",
            "entrypoint": "bash",
            "name": "gcr.io/cloud-builders/gsutil",
            "pullTiming": Object {
              "endTime": "2022-08-17T20:44:45.023016172Z",
              "startTime": "2022-08-17T20:44:45.021642025Z",
            },
            "status": "FAILURE",
            "timing": Object {
              "endTime": "2022-08-17T20:44:45.676265144Z",
              "startTime": "2022-08-17T20:44:45.021642025Z",
            },
            "volumes": Array [
              Object {
                "name": "caching.home",
                "path": "/cachinghome",
              },
            ],
            "waitFor": Array [
              "-",
            ],
          },
          Object {
            "args": Array [
              "-c",
              "set -x
export CACHING_HOME=\\"/cachinghome\\"
USER_HOME=\\"/root\\"
M2_HOME=\\"\${USER_HOME}/.m2\\"
M2_CACHE=\\"\${CACHING_HOME}/maven\\"
GRADLE_HOME=\\"\${USER_HOME}/.gradle\\"
GRADLE_CACHE=\\"\${CACHING_HOME}/gradle\\"

echo \\"Generating symbolic links for caches\\"
mkdir -p \${M2_CACHE}
mkdir -p \${GRADLE_CACHE}

[[ -d \\"\${M2_CACHE}\\" && ! -d \\"\${M2_HOME}\\" ]] && ln -s \\"\${M2_CACHE}\\" \\"\${M2_HOME}\\"
[[ -d \\"\${GRADLE_CACHE}\\" && ! -d \\"\${GRADLE_HOME}\\" ]] && ln -s \\"\${GRADLE_CACHE}\\" \\"\${GRADLE_HOME}\\"
./gradlew check",
            ],
            "entrypoint": "/bin/bash",
            "id": "test",
            "name": "openjdk:11",
            "status": "QUEUED",
            "volumes": Array [
              Object {
                "name": "caching.home",
                "path": "/cachinghome",
              },
            ],
          },
          Object {
            "args": Array [
              "-c",
              "set -x
export CACHING_HOME=\\"/cachinghome\\"
USER_HOME=\\"/root\\"
M2_HOME=\\"\${USER_HOME}/.m2\\"
M2_CACHE=\\"\${CACHING_HOME}/maven\\"
GRADLE_HOME=\\"\${USER_HOME}/.gradle\\"
GRADLE_CACHE=\\"\${CACHING_HOME}/gradle\\"

echo \\"Generating symbolic links for caches\\"
mkdir -p \${M2_CACHE}
mkdir -p \${GRADLE_CACHE}

[[ -d \\"\${M2_CACHE}\\" && ! -d \\"\${M2_HOME}\\" ]] && ln -s \\"\${M2_CACHE}\\" \\"\${M2_HOME}\\"
[[ -d \\"\${GRADLE_CACHE}\\" && ! -d \\"\${GRADLE_HOME}\\" ]] && ln -s \\"\${GRADLE_CACHE}\\" \\"\${GRADLE_HOME}\\"
./gradlew jib --image=gcr.io/j1-gc-integration-dev-v3/hello-cloud-build:",
            ],
            "entrypoint": "/bin/bash",
            "id": "build-image",
            "name": "openjdk:11",
            "status": "QUEUED",
            "volumes": Array [
              Object {
                "name": "caching.home",
                "path": "/cachinghome",
              },
            ],
          },
          Object {
            "args": Array [
              "run",
              "deploy",
              "--image=gcr.io/j1-gc-integration-dev-v3/hello-cloud-build:",
              "--platform=managed",
              "--project=j1-gc-integration-dev-v3",
              "--region=us-central1",
              "--allow-unauthenticated",
              "--memory=256Mi",
              "--set-env-vars=SPRING_PROFILES_ACTIVE=gcp",
              "hello-cloud-build",
            ],
            "id": "deploy",
            "name": "gcr.io/cloud-builders/gcloud",
            "status": "QUEUED",
            "volumes": Array [
              Object {
                "name": "caching.home",
                "path": "/cachinghome",
              },
            ],
          },
          Object {
            "args": Array [
              "-c",
              "tar -czf /tmp/gradle-cache.tar.gz gradle/ &&
gsutil cp /tmp/gradle-cache.tar.gz gs://<some bucket for caching gradle file>/<file to hold cached content>.tar.gz
",
            ],
            "dir": "/cachinghome",
            "entrypoint": "bash",
            "name": "gcr.io/cloud-builders/gsutil",
            "status": "QUEUED",
            "volumes": Array [
              Object {
                "name": "caching.home",
                "path": "/cachinghome",
              },
            ],
            "waitFor": Array [
              "deploy",
            ],
          },
        ],
        "substitutions": Object {
          "_GCS_CACHE_BUCKET": "<some bucket for caching gradle file>",
          "_GCS_CACHE_FILE": "<file to hold cached content>",
        },
        "timeout": "600s",
        "timing": Object {
          "BUILD": Object {
            "endTime": "2022-08-17T20:44:45.676362304Z",
            "startTime": "2022-08-17T20:44:44.385497769Z",
          },
          "FETCHSOURCE": Object {
            "endTime": "2022-08-17T20:44:44.385391800Z",
            "startTime": "2022-08-17T20:44:38.927319156Z",
          },
        },
      },
    },
  ],
  "_type": "google_cloud_build",
  "active": undefined,
  "createTime": 1660769077573,
  "createdOn": undefined,
  "displayName": "projects/167984947943/locations/global/builds/58991a72-f158-481d-ad66-771b90ce3e65",
  "finishTime": 1660769086243,
  "id": "58991a72-f158-481d-ad66-771b90ce3e65",
  "logUrl": "https://console.cloud.google.com/cloud-build/builds/58991a72-f158-481d-ad66-771b90ce3e65?project=167984947943",
  "logsBucket": "gs://167984947943.cloudbuild-logs.googleusercontent.com",
  "name": "projects/167984947943/locations/global/builds/58991a72-f158-481d-ad66-771b90ce3e65",
  "options.logging": "LEGACY",
  "options.pool": undefined,
  "projectId": "j1-gc-integration-dev-v3",
  "queueTtl": "3600s",
  "results.buildStepImages": Array [
    "sha256:0229d2fde1be1a278c6bb73fbfd98b2df63c30cb9012c882bdfa228e738afdf5",
    "",
    "",
    "",
    "",
  ],
  "results.buildStepOutputs": Array [
    "",
    "",
    "",
    "",
    "",
  ],
  "source.storageSource.bucket": "j1-gc-integration-dev-v3_cloudbuild",
  "source.storageSource.generation": "1660769077096260",
  "source.storageSource.object": "source/1660769074.865182-d160c8803b704c60b50e68d2b3b97d73.tgz",
  "startTime": 1660769078421,
  "status": "FAILURE",
  "timeout": "600s",
}
`;

exports[`#cloud-build entities google_cloud_build_trigger 1`] = `
Object {
  "_class": Array [
    "Rule",
  ],
  "_key": "0f5d9ec8-f7d7-4067-8912-baf2133be725",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "createTime": "2022-08-24T22:09:57.281610382Z",
        "gitFileSource": Object {
          "path": "cloudbuild.yaml",
          "repoType": "GITHUB",
          "revision": "refs/heads/master",
          "uri": "https://github.com/sample-user/sample-repo",
        },
        "id": "0f5d9ec8-f7d7-4067-8912-baf2133be725",
        "name": "testing",
        "sourceToBuild": Object {
          "ref": "refs/heads/master",
          "repoType": "GITHUB",
          "uri": "https://github.com/sample-user/sample-repo",
        },
      },
    },
  ],
  "_type": "google_cloud_build_trigger",
  "createdOn": undefined,
  "displayName": "testing",
  "id": "0f5d9ec8-f7d7-4067-8912-baf2133be725",
  "name": "testing",
}
`;
